README.md
# yang

This role provides the foundation for building network roles by providing
modules and plugins related to yang support.

To install this role: `ansible-galaxy install ansible-network.yang`

To see the version of this role you currently have installed: `ansible-galaxy list | grep yang`

To ensure you have the latest version available: `ansible-galaxy install -f ansible-network.yang`

To use this role, follow the [User Guide](https://github.com/ansible-network/yang/blob/devel/docs/user_guide/README.md).

To find other roles maintained by the Ansible Network team, see our [Galaxy Profile](https://galaxy.ansible.com/ansible-network/). 

Any open bugs and/or feature requests are tracked in [GitHub issues](https://github.com/ansible-network/yang/issues).

Interested in contributing to this role? Check out [CONTRIBUTING](https://github.com/ansible-network/yang/blob/devel/CONTRIBUTING.md) before submitting a pull request.

## Documentation

* User guide: [How to use](https://github.com/ansible-network/yang/blob/devel/docs/user_guide/README.md)
* Test guide: [How to test](https://github.com/ansible-network/yang/blob/devel/docs/tests/test_guide.md)

For module documentation see the [modules](#modules) section below.

## Requirements

* Ansible 2.6.0 or later
* Ansible Network Engine Role 2.6.2 or later

## List of network os the role is actively tested against
* iosxr (version 6.1.2)
* junos (version 17.4)

## Functions

This section provides a list of the available functions that are including in this role.
Any of the provided functions can be implemented in Ansible playbooks to perform automation activities
on yang/netconf supported devices.

* `configure` [source](https://github.com/ansible-network/yang/blob/devel/tasks/configure.yml) [docs](https://github.com/ansible-network/yang/blob/devel/docs/tasks/configure.md).
* `spec` [source](https://github.com/ansible-network/yang/blob/devel/tasks/spec.yml) [docs](https://github.com/ansible-network/yang/blob/devel/docs/tasks/spec.md).
* `fetch` [source](https://github.com/ansible-network/yang/blob/devel/tasks/fetch.yml) [docs](https://github.com/ansible-network/yang/blob/devel/docs/tasks/fetch.md).
* `get` [source](https://github.com/ansible-network/yang/blob/devel/tasks/get.yml) [docs](https://github.com/ansible-network/yang/blob/devel/docs/tasks/get.md).

## Variables

The following are the list of variables for each of the role functions.

* `configure`: [options](https://github.com/ansible-network/yang/blob/devel/meta/configure_options.yml)
* `spec`: [options](https://github.com/ansible-network/yang/blob/devel/meta/spec_options.yml)
* `fetch`: [options](https://github.com/ansible-network/yang/blob/devel/meta/fetch_options.yml)
* `fetch`: [options](https://github.com/ansible-network/yang/blob/devel/meta/get_options.yml)


## Modules

The following is a list of modules that are provided by this role, which include documentation & examples:

* `yang_fetch` [source](https://github.com/ansible-network/yang/blob/devel/action_plugins/yang_fetch.py) [docs](https://github.com/ansible-network/yang/blob/devel/library/yang_fetch.py).

## Plugins

The following is a list of plugins that are provided by this role.

### Lookup

* `yang_json2xml` [source](https://github.com/ansible-network/yang/blob/devel/lookup_plugins/yang_json2xml.py).
* `yang_spec` [source](https://github.com/ansible-network/yang/blob/devel/lookup_plugins/yang_spec.py).
* `yang_xml2json [source](https://github.com/ansible-network/yang/blob/devel/lookup_plugins/yang_xml2json.py).

### netconf

* `iosxr` [source](https://github.com/ansible-network/yang/blob/devel/netconf_plugins/iosxr.py).

### Note:
```
The iosxr netconf plugin is added to this role due the existing issue in iosxr plugin
shipped with ansible package. This issue will be fixed in ansible version 2.8
and the plugin will be removed from this role after ansible 2.8 is released.
This plugin can be used by setting configuration variable in ansible configuration file

[defaults]
netconf_plugins= <yang_role_path>/netconf_plugins

or by setting enviornment variable
$ export ANSIBLE_NETCONF_PLUGINS=<yang_role_path>/netconf_plugins
```
### Filter

## Dependencies

The following is the list of dependencies on other roles this role requires.
* Platform specific provider role task to enable Netconf on remote host

## License

GPLv3

## Author Information

Ansible Network Engineering Team


CHANGELOG.rst
====================
Ansible Network Yang
====================

.. _Ansible Network Yang_v2.7.1:

v2.7.1
======

.. _Ansible Network Yang_v2.7.1_Bugfixes:

Bugfixes
--------

- Fix issue in spec generation `yang#40 <https://github.com/ansible-network/yang/pull/40>`_.


.. _Ansible Network Yang_v2.7.0:

v2.7.0
======

.. _Ansible Network Yang_v2.7.0_New Functions:

New Functions
-------------

- Add support for get function `yang#33 <https://github.com/ansible-network/yang/pull/33>`_.


.. _Ansible Network Yang_v2.7.0_Bugfixes:

Bugfixes
--------

- Fix connection type check issue `yang#32 <https://github.com/ansible-network/yang/pull/32>`_.


.. _Ansible Network Yang_v2.6.2:

v2.6.2
======

.. _Ansible Network Yang_v2.6.2_Minor Changes:

Minor Changes
-------------

- Remove travis-ci support  `yang#30 <https://github.com/ansible-network/yang/pull/30>`_.

- Resync tox.ini file `yang#29 <https://github.com/ansible-network/yang/pull/29>`_.


.. _Ansible Network Yang_v2.6.1:

v2.6.1
======

.. _Ansible Network Yang_v2.6.1_Minor Changes:

Minor Changes
-------------

- Add network-engine role dependency in meta file `yang#14 <https://github.com/ansible-network/yang/pull/14>`_.

- Check for ansible-min-version `yang#16 <https://github.com/ansible-network/yang/pull/16>`_.

- Add support for bindep.txt `yang#25 <https://github.com/ansible-network/yang/pull/25>`_.


.. _Ansible Network Yang_v2.6.0:

v2.6.0
======

.. _Ansible Network Yang_v2.6.0_Major Changes:

Major Changes
-------------

- Initial release of the ``yang`` Ansible role.

- This role provides functions to perform automation activities using yang models using supported transports.


.. _Ansible Network Yang_v2.6.0_New Lookup Plugins:

New Lookup Plugins
------------------

- NEW ``yang_json2xml`` Plugin to validate json configuration against a yang mode and convert it to xml format.

- NEW ``yang_spec`` Plugin to generate skeleton json and xml configuration for reference config generation and yang tree representation.


.. _Ansible Network Yang_v2.6.0_New Modules:

New Modules
-----------

- NEW ``yang_fetch`` module to fetch yang models from remote network device if supported.


.. _Ansible Network Yang_v2.6.0_New Functions:

New Functions
-------------

- NEW ``fetch`` function to fetch yang models from network device if supported.

- NEW ``configure`` function to validate input ``json`` configuration against a yang model push it to device.

- NEW ``spec`` function to generate skeleton json and xml configuration for reference config generation and yang tree representation from given yang model.



lookup_plugins/yang_spec.py
# (c) 2018 Red Hat, Inc.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
#
# You should have received a copy of the GNU General Public License
# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

DOCUMENTATION = """
    lookup: yang_spec
    author: Ansible Network
    version_added: "2.6"
    short_description:  This plugin reads the content of given yang document and generates json and xml
                        configuration skeleton and a tree structure of yang document.
    description:
      - This plugin parses yang document and generates json and xml configuration skeleton and a tree
        structure of yang document. The tree structure document is as per RFC 8340 which helps to consume
        the yang document along with json and xml configuration skeleton.
    options:
      _terms:
        description: The path points to the location of the top level yang module which
        is to be transformed into to Ansible spec.
        required: True
      search_path:
        description:
          - is a colon C(:) separated list of directories to search for imported yang modules
            in the yang file mentioned in C(path) option. If the value is not given it will search in
            the current directory.
      defaults:
        description:
          - This boolean flag indicates if the generated json and xml configuration schema should have
            fields initialized with default values or not.
        default: False
      doctype:
        description:
          - Identifies the root node of the configuration skeleton. If value is C(config) only configuration
            data will be present in skeleton, if value is C(data) both config and state data fields will be present
            in output.
        default: config
        choices: ['config', 'data']
      annotations:
        description:
          - The boolean flag identifies if the xml skeleton should have comments describing the field or not.
        default: False
        type: bool
      keep_tmp_files:
        description:
          - This is a boolean flag to indicate if the intermediate files generated while creating spec
            should be kept or deleted. If the value is C(true) the files will not be deleted else by
            default all the intermediate files will be deleted irrespective of whether task run is
            successful or not. The intermediate files are stored in path C(~/.ansible/tmp/yang_spec), this
            option is mainly used for debugging purpose.
        default: False
        type: bool
"""

EXAMPLES = """
- name: Get interface yang config spec without defaults
  set_fact:
    interfaces_spec: "{{ lookup('yang_spec', 'openconfig/public/release/models/interfaces/openconfig-interfaces.yang',
                            search_path='openconfig/public/release/models:pyang/modules/', defaults=True,
                            doctype='data') }}"

- name: Get interface yang spec with defaults and state data
  set_fact:
    interfaces_spec: "{{ lookup('yang_spec', 'openconfig/public/release/models/interfaces/openconfig-interfaces.yang',
                            search_path='openconfig/public/release/models:pyang/modules/', defaults=True,
                            doctype='data') }}"
"""

RETURN = """
  _list:
    description:
      - It returns json skeleton configuration schema, xml skeleton schema and tree structure (as per RFC 8340)
        for given yang schema.
    type: complex
    contains:
      tree:
        description: The tree representation of yang scehma as per RFC 8340
        returned: success
        type: dict
        sample: |
            module: openconfig-interfaces
              +--rw interfaces
                 +--rw interface* [name]
                    +--rw name             -> ../config/name
                    +--rw config
                    |  +--rw name?            string
                    |  +--rw type             identityref
                    |  +--rw mtu?             uint16
                    |  +--rw loopback-mode?   boolean
                    |  +--rw description?     string
                    |  +--rw enabled?         boolean
                    +--ro state
                    |  +--ro name?            string
                    |  +--ro type             identityref
                    |  +--ro mtu?             uint16
                    |  +--ro loopback-mode?   boolean
                    |  +--ro description?     string
                    |  +--ro enabled?         boolean
                    |  +--ro ifindex?         uint32
                    |  +--ro admin-status     enumeration
                    |  +--ro oper-status      enumeration
                    |  +--ro last-change?     oc-types:timeticks64
      json_skeleton:
        description: The json configuration skeleton generated from yang document
        returned: success
        type: dict
        sample: |
            {
                "openconfig-interfaces:interfaces": {
                    "interface": [
                        {
                            "hold-time": {
                                "config": {
                                    "down": "",
                                    "up": ""
                                }
                            },
                            "config": {
                                "description": "",
                                "type": "",
                                "enabled": "",
                                "mtu": "",
                                "loopback-mode": "",
                                "name": ""
                            },
                            "name": "",
                            "subinterfaces": {
                                "subinterface": [
                                    {
                                        "index": "",
                                        "config": {
                                            "index": "",
                                            "enabled": "",
                                            "description": ""
                                        }
                                    }
                                ]
                            }
                        }
                    ]
                }
      xml_skeleton:
        description: The xml configuration skeleton generated from yang document
        returned: success
        type: dict
        sample: |
            <config xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
              <interfaces xmlns="http://openconfig.net/yang/interfaces">
                <interface>
                  <name/>
                  <config>
                    <name/>
                    <type/>
                    <mtu/>
                    <loopback-mode></loopback-mode>
                    <description/>
                    <enabled>True</enabled>
                  </config>
                  <hold-time>
                    <config>
                      <up></up>
                      <down></down>
                    </config>
                  </hold-time>
                  <subinterfaces>
                    <subinterface>
                      <index/>
                      <config>
                        <index></index>
                        <description/>
                        <enabled></enabled>
                      </config>
                    </subinterface>
                  </subinterfaces>
                </interface>
              </interfaces>
            </config>
"""

import os
import optparse
import sys
import shutil
import json
import uuid

import subprocess
from copy import deepcopy
from collections import Sequence

from ansible.plugins.lookup import LookupBase
from ansible.module_utils.six import StringIO
from ansible.utils.path import unfrackpath, makedirs_safe
from ansible.errors import AnsibleError

try:
    from __main__ import display
except ImportError:
    from ansible.utils.display import Display
    display = Display()

try:
    from pyang import plugin, error
except ImportError:
    raise AnsibleError("pyang is not installed")

YANG_SPEC_DIR_PATH = "~/.ansible/tmp/yang_spec"


class LookupModule(LookupBase):

    def run(self, terms, variables, **kwargs):

        res = []
        output = {}
        try:
            yang_file = terms[0]
        except IndexError:
            raise AnsibleError('the yang file must be specified')

        yang_file = os.path.realpath(os.path.expanduser(yang_file))
        if not os.path.isfile(yang_file):
            raise AnsibleError('%s invalid file path' % yang_file)

        search_path = kwargs.pop('search_path', '')
        annotations = kwargs.pop('annotations', '')

        for path in search_path.split(':'):
            path = os.path.realpath(os.path.expanduser(path))
            if path is not '' and not os.path.isdir(path):
                raise AnsibleError('%s is invalid directory path' % path)

        keep_tmp_files = kwargs.pop('keep_tmp_files', False)
        defaults = kwargs.pop('defaults', False)
        doctype = kwargs.pop('doctype', 'config')

        valid_doctype = ['config', 'data']
        if doctype not in valid_doctype:
            raise AnsibleError('doctpe value %s is invalid, valid value are %s' % (path, ', '.join(valid_doctype)))

        pyang_exec_path = find_file_in_path('pyang')

        saved_arg = deepcopy(sys.argv)
        sys.stdout = sys.stderr = StringIO()

        plugindir = unfrackpath(YANG_SPEC_DIR_PATH)
        makedirs_safe(plugindir)

        tree_file_path = os.path.join(YANG_SPEC_DIR_PATH, '%s.%s' % (str(uuid.uuid4()), 'txt'))
        xml_file_path = os.path.join(YANG_SPEC_DIR_PATH, '%s.%s' % (str(uuid.uuid4()), 'xml'))
        json_file_path = os.path.join(YANG_SPEC_DIR_PATH, '%s.%s' % (str(uuid.uuid4()), 'json'))
        tree_file_path = os.path.realpath(os.path.expanduser(tree_file_path))
        xml_file_path = os.path.realpath(os.path.expanduser(xml_file_path))
        json_file_path = os.path.realpath(os.path.expanduser(json_file_path))

        # fill in the sys args before invoking pyang to retrieve xml skeleton
        sample_xml_skeleton_cmd = [pyang_exec_path, '-f', 'sample-xml-skeleton', '-o', xml_file_path, yang_file, '-p',
                                   search_path, "--sample-xml-skeleton-doctype", doctype, "--lax-quote-checks"]

        if defaults:
            sample_xml_skeleton_cmd.append("--sample-xml-skeleton-defaults")

        if annotations:
            sample_xml_skeleton_cmd.append("--sample-xml-skeleton-annotations")

        try:
            subprocess.check_output(' '.join(sample_xml_skeleton_cmd), stderr=subprocess.STDOUT, shell=True)
        except SystemExit:
            pass
        except Exception as e:
            if not keep_tmp_files:
                shutil.rmtree(os.path.realpath(os.path.expanduser(YANG_SPEC_DIR_PATH)), ignore_errors=True)
            raise AnsibleError('Error while generating skeleton xml file: %s' % e)
        finally:
            err = sys.stdout.getvalue()
            if err and 'error' in err.lower():
                if not keep_tmp_files:
                    shutil.rmtree(os.path.realpath(os.path.expanduser(YANG_SPEC_DIR_PATH)), ignore_errors=True)
                raise AnsibleError('Error while generating skeleton xml file: %s' % err)

        sys.stdout.flush()
        sys.stderr.flush()

        # fill in the sys args before invoking pyang to retrieve tree structure
        tree_cmd = [pyang_exec_path, '-f', 'tree', '-o', tree_file_path, yang_file, '-p', search_path,
                    "--lax-quote-checks"]

        try:
            subprocess.check_output(' '.join(tree_cmd), stderr=subprocess.STDOUT, shell=True)
        except SystemExit:
            pass
        except Exception as e:
            if not keep_tmp_files:
                shutil.rmtree(os.path.realpath(os.path.expanduser(YANG_SPEC_DIR_PATH)), ignore_errors=True)
            raise AnsibleError('Error while generating tree file: %s' % e)
        finally:
            err = sys.stdout.getvalue()
            if err and 'error' in err.lower():
                if not keep_tmp_files:
                    shutil.rmtree(os.path.realpath(os.path.expanduser(YANG_SPEC_DIR_PATH)), ignore_errors=True)
                raise AnsibleError('Error while generating tree file: %s' % err)

        sys.stdout.flush()
        sys.stderr.flush()

        plugin_file_src = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'yang_spec.py')
        shutil.copy(plugin_file_src, plugindir)

        # fill in the sys args before invoking pyang to retrieve json skeleton
        sample_json_skeleton_cmd = [pyang_exec_path, '--plugindir', plugindir, '-f', 'sample-json-skeleton', '-o',
                                    json_file_path, yang_file, '-p', search_path, '--lax-quote-checks',
                                    '--sample-json-skeleton-doctype', doctype]

        if defaults:
            sample_json_skeleton_cmd.append("--sample-json-skeleton-defaults")

        try:
            subprocess.check_output(' '.join(sample_json_skeleton_cmd), stderr=subprocess.STDOUT, shell=True)
        except SystemExit:
            pass
        except Exception as e:
            if not keep_tmp_files:
                shutil.rmtree(os.path.realpath(os.path.expanduser(YANG_SPEC_DIR_PATH)), ignore_errors=True)
            raise AnsibleError('Error while generating skeleton json file: %s' % e)
        finally:
            err = sys.stdout.getvalue()
            if err and 'error' in err.lower():
                if not keep_tmp_files:
                    shutil.rmtree(os.path.realpath(os.path.expanduser(YANG_SPEC_DIR_PATH)), ignore_errors=True)
                raise AnsibleError('Error while generating tree json: %s' % err)

        with open(tree_file_path, 'r') as f:
            output['tree'] = f.read()

        with open(xml_file_path, 'r') as f:
            output['xml_skeleton'] = f.read()

        with open(json_file_path, 'r') as f:
            output['json_skeleton'] = json.load(f)

        if not keep_tmp_files:
            shutil.rmtree(plugindir, ignore_errors=True)
        res.append(output)

        sys.argv = saved_arg
        return res


def find_file_in_path(filename):
    # Check $PATH first, followed by same directory as sys.argv[0]
    paths = os.environ['PATH'].split(os.pathsep) + [os.path.dirname(sys.argv[0])]
    for dirname in paths:
        fullpath = os.path.join(dirname, filename)
        if os.path.isfile(fullpath):
            return fullpath


def pyang_plugin_init():
    plugin.register_plugin(SampleJSONSkeletonPlugin())


def to_list(val):
    if isinstance(val, Sequence):
        return list(val)
    elif val is not None:
        return [val]
    else:
        return list()


class SampleJSONSkeletonPlugin(plugin.PyangPlugin):

    def add_opts(self, optparser):
        optlist = [
            optparse.make_option("--sample-json-skeleton-doctype",
                                 dest="doctype",
                                 default="data",
                                 help="Type of sample JSON document (data or config)."),
            optparse.make_option("--sample-json-skeleton-defaults",
                                 action="store_true",
                                 dest="sample_defaults",
                                 default=False,
                                 help="Insert data with defaults values."),
        ]
        g = optparser.add_option_group(
            "Sample-json-skeleton output specific options")
        g.add_options(optlist)

    def add_output_format(self, fmts):
        self.multiple_modules = True
        fmts['sample-json-skeleton'] = self

    def setup_fmt(self, ctx):
        ctx.implicit_errors = False

    def emit(self, ctx, modules, fd):
        """Main control function.
        """
        for (epos, etag, eargs) in ctx.errors:
            if error.is_error(error.err_level(etag)):
                raise error.EmitError("sample-json-skeleton plugin needs a valid module")
        tree = {}
        self.defaults = ctx.opts.sample_defaults
        self.doctype = ctx.opts.doctype
        if self.doctype not in ("config", "data"):
            raise error.EmitError("Unsupported document type: %s" %
                                  self.doctype)

        for module in modules:
            self.process_children(module, tree, None)
        json.dump(tree, fd, indent=4)

    def process_children(self, node, parent, pmod):
        """Process all children of `node`, except "rpc" and "notification".
        """
        for ch in node.i_children:
            if self.doctype == 'config' and not ch.i_config:
                continue
            if ch.keyword in ["rpc", "notification"]:
                continue
            if ch.keyword in ["choice", "case"]:
                self.process_children(ch, parent, pmod)
                continue
            if ch.i_module.i_modulename == pmod:
                nmod = pmod
                nodename = ch.arg
            else:
                nmod = ch.i_module.i_modulename
                nodename = "%s:%s" % (nmod, ch.arg)
            if ch.keyword == "container":
                ndata = dict()
                self.process_children(ch, ndata, nmod)
                parent[nodename] = ndata
            elif ch.keyword == "list":
                ndata = list()
                ndata.append({})
                self.process_children(ch, ndata[0], nmod)
                parent[nodename] = ndata
            elif ch.keyword == "leaf":
                if ch.arg == 'keepalive-interval':
                    pass
                ndata = str(ch.i_default) if (self.defaults and ch.i_default is not None) else ""
            elif ch.keyword == "leaf-list":
                ndata = to_list(str(ch.i_default)) if (self.defaults and ch.i_default is not None) else [""]
            parent[nodename] = ndata

    def base_type(self, type):
        """Return the base type of `type`."""
        while 1:
            if type.arg == "leafref":
                node = type.i_type_spec.i_target_node
            elif type.i_typedef is None:
                break
            else:
                node = type.i_typedef
            type = node.search_one("type")
        if type.arg == "decimal64":
            return [type.arg, int(type.search_one("fraction-digits").arg)]
        elif type.arg == "union":
            return [type.arg, [self.base_type(x) for x in type.i_type_spec.types]]
        else:
            return type.arg


tools/test-setup.sh
#!/bin/bash -ex
# Copyright 2018 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# NOTE(pabelanger): Create a symlink for ansible-network.network-engine to keep
# meta/main.yml happy
ln -s ~/src/github.com/ansible-network/network-engine \
  ~/src/github.com/ansible-network/ansible-network.network-engine


lookup_plugins/yang_json2xml.py
# (c) 2018 Red Hat, Inc.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
#
# You should have received a copy of the GNU General Public License
# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

DOCUMENTATION = """
lookup: yang_json2xml
author: Ansible Network
version_added: "2.6"
short_description: Validates json configuration against yang data model and convert it to xml.
description:
  - This plugin lookups the input json configuration, validates it against the respective yang data
    model which is also given as input to this plugin and coverts it to xml format which can be used
    as payload within Netconf rpc.
options:
  _terms:
    description:
      - Input json configuration file path that adheres to a particular yang model.
    required: True
    type: path
  doctype:
    description:
      - Specifies the target root node of the generated xml. The default value is C(config)
    default: config
  yang_file:
    description:
      - Path to yang model file against which the json configuration is validated and
        converted to xml.
    required: True
    type: path
  search_path:
    description:
      - This option is a colon C(:) separated list of directories to search for imported yang modules
        in the yang file mentioned in C(path) option. If the value is not given it will search in
        the current directory.
    required: false
  keep_tmp_files:
    description:
      - This is a boolean flag to indicate if the intermediate files generated while validation json
       configuration should be kept or deleted. If the value is C(true) the files will not be deleted else by
        default all the intermediate files will be deleted irrespective of whether task run is
        successful or not. The intermediate files are stored in path C(~/.ansible/tmp/json2xml), this
        option is mainly used for debugging purpose.
    default: False
    type: bool
"""

EXAMPLES = """
- name: translate json to xml
  debug: msg="{{ lookup('yang_json2xml', config_json,
                         yang_file='openconfig/public/release/models/interfaces/openconfig-interfaces.yang',
                         search_path='openconfig/public/release/models:pyang/modules/') }}"
"""

RETURN = """
_raw:
   description: The translated xml string from json
"""

import os
import imp
import re
import sys
import json
import shutil
import uuid
import glob

from copy import deepcopy

from ansible.plugins.lookup import LookupBase
from ansible.module_utils.six import StringIO
from ansible.utils.path import unfrackpath, makedirs_safe
from ansible.module_utils._text import to_text
from ansible.errors import AnsibleError

try:
    import pyang        # noqa
except ImportError:
    raise AnsibleError("pyang is not installed")

try:
    from lxml import etree
except ImportError:
    raise AnsibleError("lxml is not installed")

try:
    from __main__ import display
except ImportError:
    from ansible.utils.display import Display

    display = Display()

JSON2XML_DIR_PATH = "~/.ansible/tmp/json2xml"


class LookupModule(LookupBase):

    def run(self, terms, variables, **kwargs):

        res = []
        try:
            json_config = terms[0]
        except IndexError:
            raise AnsibleError("path to json file must be specified")

        try:
            yang_file = kwargs['yang_file']
        except KeyError:
            raise AnsibleError("value of 'yang_file' must be specified")

        yang_file = os.path.realpath(os.path.expanduser(yang_file))
        if not os.path.isfile(yang_file):
            # Maybe we are passing a glob?
            yang_files = glob.glob(yang_file)
            if not yang_files:
                # Glob returned no files
                raise AnsibleError('%s invalid file path' % yang_file)
        else:
            yang_files = [yang_file]

        search_path = kwargs.pop('search_path', '')
        keep_tmp_files = kwargs.pop('keep_tmp_files', False)

        abs_search_path = None
        for path in search_path.split(':'):
            path = os.path.realpath(os.path.expanduser(path))
            if abs_search_path is None:
                abs_search_path = path
            else:
                abs_search_path += ':' + path
            if path is not '' and not os.path.isdir(path):
                raise AnsibleError('%s is invalid directory path' % path)

        search_path = abs_search_path
        json_config = os.path.realpath(os.path.expanduser(json_config))
        try:
            # validate json
            with open(json_config) as fp:
                json.load(fp)
        except Exception as exc:
            raise AnsibleError("Failed to load json configuration: %s" % (to_text(exc, errors='surrogate_or_strict')))

        root_node = kwargs.get('doctype', 'config')

        base_pyang_path = sys.modules['pyang'].__file__
        pyang_exec_path = find_file_in_path('pyang')
        pyang_exec = imp.load_source('pyang', pyang_exec_path)

        saved_arg = deepcopy(sys.argv)
        sys.modules['pyang'].__file__ = base_pyang_path

        saved_stdout = sys.stdout
        saved_stderr = sys.stderr
        sys.stdout = sys.stderr = StringIO()

        plugindir = unfrackpath(JSON2XML_DIR_PATH)
        makedirs_safe(plugindir)

        jtox_file_path = os.path.join(JSON2XML_DIR_PATH, '%s.%s' % (str(uuid.uuid4()), 'jtox'))
        xml_file_path = os.path.join(JSON2XML_DIR_PATH, '%s.%s' % (str(uuid.uuid4()), 'xml'))
        jtox_file_path = os.path.realpath(os.path.expanduser(jtox_file_path))
        xml_file_path = os.path.realpath(os.path.expanduser(xml_file_path))

        yang_metada_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), '../files/yang')
        yang_metadata_path = os.path.join(yang_metada_dir, 'nc-op.yang')
        search_path += ":%s" % yang_metada_dir

        # fill in the sys args before invoking pyang
        sys.argv = [pyang_exec_path, '-f', 'jtox', '-o', jtox_file_path, '-p', search_path,
                    "--lax-quote-checks"] + yang_files + [yang_metadata_path]

        try:
            pyang_exec.run()
        except SystemExit:
            pass
        except Exception as e:
            shutil.rmtree(os.path.realpath(os.path.expanduser(JSON2XML_DIR_PATH)), ignore_errors=True)
            raise AnsibleError('Error while generating intermediate (jtox) file: %s' % e)
        finally:
            err = sys.stderr.getvalue()
            if err and 'error' in err.lower():
                if not keep_tmp_files:
                    shutil.rmtree(os.path.realpath(os.path.expanduser(JSON2XML_DIR_PATH)), ignore_errors=True)
                raise AnsibleError('Error while generating intermediate (jtox) file: %s' % err)

        json2xml_exec_path = find_file_in_path('json2xml')
        json2xml_exec = imp.load_source('json2xml', json2xml_exec_path)

        # fill in the sys args before invoking json2xml
        sys.argv = [json2xml_exec_path, '-t', root_node, '-o', xml_file_path, jtox_file_path, json_config]

        try:
            json2xml_exec.main()
            with open(xml_file_path, 'r+') as fp:
                content = fp.read()

        except SystemExit:
            pass
        finally:
            err = sys.stderr.getvalue()
            if err and 'error' in err.lower():
                if not keep_tmp_files:
                    shutil.rmtree(os.path.realpath(os.path.expanduser(JSON2XML_DIR_PATH)), ignore_errors=True)
                raise AnsibleError('Error while translating to xml: %s' % err)
            sys.argv = saved_arg
            sys.stdout = saved_stdout
            sys.stderr = saved_stderr

        try:
            content = re.sub(r'<\? ?xml .*\? ?>', '', content)
            root = etree.fromstring(content)
        except Exception as e:
            raise AnsibleError('Error while reading xml document: %s' % e)
        finally:
            if not keep_tmp_files:
                shutil.rmtree(os.path.realpath(os.path.expanduser(JSON2XML_DIR_PATH)), ignore_errors=True)
        res.append(etree.tostring(root))

        return res


def find_file_in_path(filename):
    # Check $PATH first, followed by same directory as sys.argv[0]
    paths = os.environ['PATH'].split(os.pathsep) + [os.path.dirname(sys.argv[0])]
    for dirname in paths:
        fullpath = os.path.join(dirname, filename)
        if os.path.isfile(fullpath):
            return fullpath
